server/datastore/mysql/host_certificates.go

```diff
@@ -0,0 +1,197 @@
+
+func (ds *Datastore) ListHostCertificates(ctx context.Context, hostID uint, opts fleet.ListOptions) ([]*fleet.HostCertificateRecord, *fleet.PaginationMetadata, error) {
+	return listHostCertsDB(ctx, ds.reader(ctx), hostID, opts)
+}
+
+func (ds *Datastore) UpdateHostCertificates(ctx context.Context, hostID uint, certs []*fleet.HostCertificateRecord) error {
+	incomingBySHA1 := make(map[string]*fleet.HostCertificateRecord, len(certs))
+	for _, cert := range certs {
+		if cert.HostID != hostID {
+			// caller should ensure this does not happen
+			level.Debug(ds.logger).Log("msg", fmt.Sprintf("host ID does not match provided certificate: %d %d", hostID, cert.HostID))
+		}
+		if _, ok := incomingBySHA1[strings.ToUpper(hex.EncodeToString(cert.SHA1Sum))]; ok {
+			// TODO: sha1 is broken so this could be a sign of a problem, how should we handle?
+			level.Info(ds.logger).Log("msg", "host has multiple certificates with the same SHA1, only the first will be recorded", "host_id", hostID, "sha1", string(cert.SHA1Sum))
+			continue
+		}
+		incomingBySHA1[strings.ToUpper(hex.EncodeToString(cert.SHA1Sum))] = cert
+	}
+
+	// get existing certs for this host; we'll use the reader because we expect certs to change
+	// infrequently and they will be eventually consistent
+	existingCerts, _, err := listHostCertsDB(ctx, ds.reader(ctx), hostID, fleet.ListOptions{}) // requesting unpaginated results with default limit of 1 million
+	if err != nil {
+		return fmt.Errorf("list host certs for update: %w", err)
+	}
+	existingBySHA1 := make(map[string]*fleet.HostCertificateRecord, len(existingCerts))
+	for _, ec := range existingCerts {
+		existingBySHA1[strings.ToUpper(hex.EncodeToString(ec.SHA1Sum))] = ec
+	}
+
+	toInsert := make([]*fleet.HostCertificateRecord, 0, len(incomingBySHA1))
+	// toUpdate := make([]*fleet.HostCertificateRecord, 0, len(incomingBySHA1))
+	for sha1, incoming := range incomingBySHA1 {
+		if _, ok := existingBySHA1[sha1]; ok {
+			// TODO: should we always update existing records? skipping updates reduces db load but
+			// osquery is using sha1 so we consider subtleties
+			level.Debug(ds.logger).Log("msg", fmt.Sprintf("existing certificate: %s", sha1), "host_id", hostID)
+		} else {
+			toInsert = append(toInsert, incoming)
+		}
+	}
+
+	toDelete := make([]uint, 0, len(existingBySHA1))
+	for sha1, existing := range existingBySHA1 {
+		if _, ok := incomingBySHA1[sha1]; !ok {
+			toDelete = append(toDelete, existing.ID)
+		}
+	}
+
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		if err := insertHostCertsDB(ctx, tx, toInsert); err != nil {
+			return ctxerr.Wrap(ctx, err, "insert host certs")
+		}
+		if err := softDeleteHostCertsDB(ctx, tx, hostID, toDelete); err != nil {
+			return ctxerr.Wrap(ctx, err, "soft delete host certs")
+		}
+		return nil
+	})
+}
+
+func listHostCertsDB(ctx context.Context, tx sqlx.QueryerContext, hostID uint, opts fleet.ListOptions) ([]*fleet.HostCertificateRecord, *fleet.PaginationMetadata, error) {
```

Comment:
Heads-up for a future PR, starting in this function and below, it's missing the `ctxerr.Wrap` calls around errors, which should be done in the Datastore (closest to the original error).